core/src/banking_stage.rs

```diff
@@ -180,78 +182,68 @@ impl BankingStage {
+                if bank.is_none() {
+                    rebuffered_packets_len += old_unprocessed_indexes.len();
+                    return Self::packet_has_more_unprocessed_transactions(
+                        &old_unprocessed_indexes,
+                    );
+                }
+                let bank = bank.unwrap();
 
-            new_tx_count += processed;
-
-            // Collect any unprocessed transactions in this batch for forwarding
-            rebuffered_packets += new_unprocessed_indexes.len();
-            Self::push_unprocessed(
-                &mut unprocessed_packets,
-                msgs,
-                new_unprocessed_indexes,
-                &mut dropped_batches_count,
-                batch_limit,
-            );
-
-            if processed < verified_txs_len {
-                let next_leader = poh_recorder.lock().unwrap().next_slot_leader();
-                // Walk thru rest of the transactions and filter out the invalid (e.g. too old) ones
-                #[allow(clippy::while_let_on_iterator)]
-                while let Some((msgs, unprocessed_indexes)) = buffered_packets_iter.next() {
-                    let unprocessed_indexes = Self::filter_unprocessed_packets(
+                let (processed, verified_txs_len, new_unprocessed_indexes) =
+                    Self::process_received_packets(
                         &bank,
+                        &poh_recorder,
                         &msgs,
-                        &unprocessed_indexes,
-                        my_pubkey,
-                        next_leader,
-                    );
-                    Self::push_unprocessed(
-                        &mut unprocessed_packets,
-                        msgs,
-                        unprocessed_indexes,
-                        &mut dropped_batches_count,
-                        batch_limit,
+                        old_unprocessed_indexes.to_owned(),
+                        transaction_status_sender.clone(),
+                        gossip_vote_sender,
                     );
+                new_tx_count += processed;
+
+                // Out of the buffered packets just retried, collect any still unprocessed
+                // transactions in this batch for forwarding
+                rebuffered_packets_len += new_unprocessed_indexes.len();
+                if processed < verified_txs_len {
+                    reached_end_of_slot =
+                        Some((poh_recorder.lock().unwrap().next_slot_leader(), bank));
+                }
+                let has_more_unprocessed_transactions =
+                    Self::packet_has_more_unprocessed_transactions(&new_unprocessed_indexes);
+                if has_more_unprocessed_transactions {
+                    *old_unprocessed_indexes = new_unprocessed_indexes;
                 }
+                if let Some(test_fn) = &test_fn {
+                    test_fn();
+                }
+                has_more_unprocessed_transactions
```

Comment:
Thought about this subtle difference between this and there(https://github.com/solana-labs/solana/pull/15210/files#r573986343). And I just noticed this is basically same thing? How about defining a tiny helper like this?

```rust
let should_retain = update_retained_packets_with_still_unprocessed(...)
if let Some(test_fn) = ... {   ...   }
should_retain

...

fn update_retained_packets_with_still_unprocessed(old_unprocessed_indexes: &mut Vec<usize>, new_unprocessed_indexes: _) -> bool {
  let should_retain = Self::packet_has_more_unprocessed_transactions(&new_unprocessed_indexes);
  if should_retain {
    *old_unprocessed_indexes = new_unprocessed_indexes
  };
  should_retain;
}
```

Reply:
Updated